crates/ethereum-forks/src/hardforks/mod.rs

```diff
@@ -0,0 +1,124 @@
+/// Ethereum helper methods
+mod ethereum;
+pub use ethereum::{EthereumActivations, EthereumHardforks};
+
+/// Optimism helper methods
+#[cfg(feature = "optimism")]
+mod optimism;
+#[cfg(feature = "optimism")]
+pub use optimism::{OptimismActivations, OptimismHardforks};
+
+use crate::{ForkCondition, Hardfork};
+use std::collections::HashMap;
+
+/// Generic trait over a set of ordered hardforks
+pub trait Hardforks: Default + Clone {
+    /// Retrieves [`ForkCondition`] from `fork`. If `fork` is not present, returns
+    /// [`ForkCondition::Never`].
+    fn fork<H: Hardfork>(&self, fork: H) -> ForkCondition;
+
+    /// Get an iterator of all hardforks with their respective activation conditions.
+    fn forks_iter(&self) -> impl Iterator<Item = (&Box<dyn Hardfork + 'static>, ForkCondition)>;
+
+    /// Convenience method to check if a fork is active at a given timestamp.
+    fn is_fork_active_at_timestamp<H: Hardfork>(&self, fork: H, timestamp: u64) -> bool {
+        self.fork(fork).active_at_timestamp(timestamp)
+    }
+
+    /// Convenience method to check if a fork is active at a given block number.
+    fn is_fork_active_at_block<H: Hardfork>(&self, fork: H, block_number: u64) -> bool {
+        self.fork(fork).active_at_block(block_number)
+    }
+}
+
+/// Ordered list of a chain hardforks that implement [`Hardfork`].
+#[derive(Default, Clone, PartialEq, Eq)]
+pub struct ChainHardforks {
+    forks: Vec<(Box<dyn Hardfork>, ForkCondition)>,
+    map: HashMap<&'static str, ForkCondition>,
+}
+
+impl ChainHardforks {
+    /// Creates a new [`ChainHardforks`] from a list.
+    pub fn new(forks: Vec<(Box<dyn Hardfork>, ForkCondition)>) -> Self {
```

Comment:
this should mention that the caller must ensure that these are ordered by activation.

and it is expected that these include equivalent ethereum hardforks.